### PR TITLE
[Sparse] Update code and add unittest for `formats` 

### PR DESCRIPTION
**Pull request by keli-wen**
## Description

Based on the previous [PR#5851 [Sparse] Clean formats function docstring](https://github.com/dmlc/dgl/pull/5851) and [Issue#5846: [Sparse] Standardize the behavior of `formats()`](https://github.com/dmlc/dgl/issues/5846), this PR primarily accomplishes the following:

- [x] Modified the underlying implementation of the `formats()` function within `unit_graph.cc`.
- [x] Added some new, necessary unit tests for the `formats()` function in `tests/python/common/test_heterograph-misc.py`.

With these changes, the behavior of `formats()` will be consistent with its description in `heterograph.py`.
## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [x] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [x] All changes have test coverage
- [x] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Comment by Rhett-Ying** [comment]
@dgl-bot run r6i.16xlarge bench_format_conversion dmlc/master

**Comment by Rhett-Ying** [comment]
@dgl-bot run g4dn.16xlarge bench_format_conversion dmlc/master

File: src/graph/unit_graph.cc [hunk 1]
```diff
@@ -1529,18 +1529,32 @@ HeteroGraphPtr UnitGraph::GetFormat(SparseFormat format) const {
...
+  // If the intersection of formats and created_formats is not empty.
+  // The format(s) in the intersection will be retained.
+  if (intersection != 0) {
+    COOPtr coo_ptr = nullptr;
+    CSRPtr in_csr_ptr = nullptr;
+    CSRPtr out_csr_ptr = nullptr;
+
+    if (COO_CODE & intersection) coo_ptr = GetCOO(false);
+    if (CSC_CODE & intersection) in_csr_ptr = GetInCSR(false);
+    if (CSR_CODE & intersection) out_csr_ptr = GetOutCSR(false);
```
Comment (frozenbugs) [inline src/graph/unit_graph.cc:1546]: Same for others.
```suggestion
    COOPtr coo_ptr = COO_CODE & intersection ? GetCOO(false) : nullptr;
```

**Review by frozenbugs — APPROVED** [review]

**Comment by frozenbugs** [comment]
@Rhett-Ying please double check the benchmark result looks good to you.

**Comment by Rhett-Ying** [comment]
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:////Users/ruying/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel=File-List
href="file:////Users/ruying/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">



  | test_name | params | unit |   | number_5859 | machine | number_master
-- | -- | -- | -- | -- | -- | -- | --
0 | api.bench_format_conversion.track_time | 'cora', ('coo', 'csc') | s | 9% | 0.00013382 | g4dn.16xlarge-gpu | 0.00012304
1 | api.bench_format_conversion.track_time | 'cora', ('csc', 'coo') | s | 1% | 0.00024748 | g4dn.16xlarge-gpu | 0.00024497
2 | api.bench_format_conversion.track_time | 'cora', ('coo', 'csr') | s | 5% | 0.00013032 | g4dn.16xlarge-gpu | 0.00012383
3 | api.bench_format_conversion.track_time | 'cora', ('csr', 'coo') | s | 12% | 0.0002709 | g4dn.16xlarge-gpu | 0.00024218
4 | api.bench_format_conversion.track_time | 'cora', ('csr', 'csc') | s | 9% | 0.00015873 | g4dn.16xlarge-gpu | 0.00014521
5 | api.bench_format_conversion.track_time | 'cora', ('csc', 'csr') | s | 11% | 0.00016037 | g4dn.16xlarge-gpu | 0.00014513
6 | api.bench_format_conversion.track_time | 'livejournal', ('coo', 'csc') | s | 0% | 0.07718615 | g4dn.16xlarge-gpu | 0.07730999
7 | api.bench_format_conversion.track_time | 'livejournal', ('csc', 'coo') | s | 0% | 0.14916816 | g4dn.16xlarge-gpu | 0.14913475
8 | api.bench_format_conversion.track_time | 'livejournal', ('coo', 'csr') | s | 0% | 0.05643929 | g4dn.16xlarge-gpu | 0.05643291
9 | api.bench_format_conversion.track_time | 'livejournal', ('csr', 'coo') | s | 0% | 0.11298267 | g4dn.16xlarge-gpu | 0.11295137
10 | api.bench_format_conversion.track_time | 'livejournal', ('csr', 'csc') | s | 0% | 0.11628831 | g4dn.16xlarge-gpu | 0.11626382
11 | api.bench_format_conversion.track_time | 'livejournal', ('csc', 'csr') | s | 0% | 0.11520695 | g4dn.16xlarge-gpu | 0.11521112



</body>

</html>

**Comment by Rhett-Ying** [comment]
@dgl-bot run g4dn.16xlarge bench_format_conversion dmlc/master

**Comment by Rhett-Ying** [comment]
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:////Users/ruying/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel=File-List
href="file:////Users/ruying/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">



  | test_name | params | unit |   | number_master | machine | number_5859
-- | -- | -- | -- | -- | -- | -- | --
0 | api.bench_format_conversion.track_time | 'cora', ('coo', 'csc') | s | 0% | 0.00012357 | g4dn.16xlarge-gpu | 0.00012394
1 | api.bench_format_conversion.track_time | 'cora', ('csc', 'coo') | s | -1% | 0.00025006 | g4dn.16xlarge-gpu | 0.00024796
2 | api.bench_format_conversion.track_time | 'cora', ('coo', 'csr') | s | -2% | 0.00012885 | g4dn.16xlarge-gpu | 0.00012603
3 | api.bench_format_conversion.track_time | 'cora', ('csr', 'coo') | s | -1% | 0.00024654 | g4dn.16xlarge-gpu | 0.00024449
4 | api.bench_format_conversion.track_time | 'cora', ('csr', 'csc') | s | -7% | 0.00016098 | g4dn.16xlarge-gpu | 0.00014895
5 | api.bench_format_conversion.track_time | 'cora', ('csc', 'csr') | s | -10% | 0.00016242 | g4dn.16xlarge-gpu | 0.00014696
6 | api.bench_format_conversion.track_time | 'livejournal', ('coo', 'csc') | s | 0% | 0.0772455 | g4dn.16xlarge-gpu | 0.07726824
7 | api.bench_format_conversion.track_time | 'livejournal', ('csc', 'coo') | s | 0% | 0.14903008 | g4dn.16xlarge-gpu | 0.14906498
8 | api.bench_format_conversion.track_time | 'livejournal', ('coo', 'csr') | s | 0% | 0.05643348 | g4dn.16xlarge-gpu | 0.05643428
9 | api.bench_format_conversion.track_time | 'livejournal', ('csr', 'coo') | s | 0% | 0.11302251 | g4dn.16xlarge-gpu | 0.11290552
10 | api.bench_format_conversion.track_time | 'livejournal', ('csr', 'csc') | s | 0% | 0.11632094 | g4dn.16xlarge-gpu | 0.11637903
11 | api.bench_format_conversion.track_time | 'livejournal', ('csc', 'csr') | s | 0% | 0.11524528 | g4dn.16xlarge-gpu | 0.11522761



</body>

</html>

**Review by Rhett-Ying — APPROVED** [review]

**Pull request merged**